mysqlnd_azure.c

```diff
@@ -174,6 +208,77 @@ set_redirect_client_options(MYSQLND_CONN_DATA * const conn, MYSQLND_CONN_DATA *
 }
 /* }}} */
 
+/* {{{ get_redirect_info */
+static zend_bool
+get_redirect_info(const MYSQLND_CONN_DATA * const conn, char* redirect_host, char* redirect_user, unsigned int* p_ui_redirect_port)
+{
+	/**
+	* Get redirected server information contained in OK packet.
+	* Redirection string has following format:
+	* Location: mysql://redirectedHostName:redirectedPort/user=redirectedUser
+	* the minimal len is 28 bytes
+	*/
+
+    const char * msg_header = "Location: mysql://";
+    int msg_header_len = strlen(msg_header);
+
+    if (conn->last_message.l <= 27 || (strncmp(conn->last_message.s, msg_header, msg_header_len) != 0)) {
+        return FALSE;
+    }
+
+    char *cur_pos = conn->last_message.s + msg_header_len;
+    char *end = conn->last_message.s + conn->last_message.l;
+
+    char* host_begin = cur_pos, *host_end = NULL, *port_begin = NULL, *port_end = NULL, *user_begin = NULL,  *user_end = NULL;
+    zend_bool found_host = FALSE, found_port = FALSE, found_user = FALSE;
+
+    while(cur_pos < end) {
+        if(found_host == FALSE && *cur_pos == ':') {
+            host_end = cur_pos;
+            cur_pos++;
+            port_begin = cur_pos;
+            found_host = TRUE;
+        }
+
+        if(found_host == TRUE && found_port == FALSE && *cur_pos == '/') {
+            port_end = cur_pos;
+            found_port = TRUE;
+
+            cur_pos++;
+            int user_delimiter_len = strlen("user=");
+            if(end - cur_pos > user_delimiter_len && strncmp(cur_pos, "user=", user_delimiter_len)==0) {
+                user_begin = cur_pos + user_delimiter_len;
+                user_end = end;
+                found_user = TRUE;
+                break;
+            }
+        }
+
+        cur_pos++;
+    }
+
+    if(!found_host || !found_port || !found_user) {
+        return FALSE;
+    }
+
+    int host_len = host_end - host_begin;
+    int port_len = port_end - port_begin;
+    int user_len = user_end - user_begin;
+
+    if(host_len<=0 || port_len<=0 || user_len<=0 || host_len > MAX_REDIRECT_HOST_LEN || port_len > 8 || user_len > MAX_REDIRECT_USER_LEN) {
```

Comment:
Just wondering if PHP has functionality of debug trace. If yes, could we add debug trace (or error trace) for negative results or exception situations. You can image that you cannot debug on customer's machine. So, when problem happened, the only thing I can image that we can do is to ask them enable debug trace if we fail to identify the root cause. But you can have your suggestion that how to handle with such case.

Reply:
You could use [zend_fetch_debug_backtrace()](https://github.com/php/php-src/blob/php-7.3.13/Zend/zend_builtin_functions.c#L2456), and print it with [zend_print_zval_r_to_str](https://github.com/php/php-src/blob/php-7.3.13/Zend/zend.c#L487) (or similar), if a respective INI option is set (maybe use `mysqlnd.debug`).

Reply:
Thanks, I think we can do it.